### ISSUE TITLE
Communication Thread Performance Degredation

### ISSUE THREAD

**Issue by gbalke** [BetzDrive/bldc-controller#30]
Continuing off of https://github.com/BetzDrive/bldc-controller-hardware/issues/9

Not sure exactly what is causing it but communication error rates are significantly higher after commit https://github.com/BetzDrive/bldc-controller/pull/27/commits/6323cb5dfdb8e68432badc53ac54649a532e646c.

I'll be investigating this but would also appreciate any input you have @luca-della-vedova. I'll likely revert changes on a branch for now so there's a non-bugged set of code.

IMO something has been off with the fact that communication performance decays significantly with less frequent scheduling and/or random changes to the firmware. To my knowledge, there's a UART peripheral on-board that should be buffering data to be processed. When a byte or bytes are ready, we process that data. It seems as if we start missing data when our chip is busier which makes me think somehow we're polling for data instead of being something in the background? I'll need to investigate as this is prior work and I'm not super familiar with the details of how ChibiOS implements UART for the STM32.

**Comment by gbalke** [comment]
Here's some empirical results from before and after the problematic commit:

Before:
```
[ INFO] [1601538569.857704926]: Comm time dt: 0.001060, freq: 943.120154, errors: 0, error rate: 0.00%
[ INFO] [1601538570.963539055]: Comm time dt: 0.001106, freq: 904.286455, errors: 0, error rate: 0.00%
[ INFO] [1601538572.024927492]: Comm time dt: 0.001061, freq: 942.163612, errors: 0, error rate: 0.00%
[ INFO] [1601538573.155823045]: Comm time dt: 0.001131, freq: 884.256581, errors: 0, error rate: 0.00%
[ INFO] [1601538574.290199357]: Comm time dt: 0.001134, freq: 881.543247, errors: 1, error rate: 0.10%
[ INFO] [1601538574.290279570]: Expected 1 Byte(s) got 0, occured 1 times                            
[ INFO] [1601538574.290305727]: std::exception, occured 1 times                                       
[ INFO] [1601538575.387504819]: Comm time dt: 0.001097, freq: 911.321252, errors: 0, error rate: 0.00%
[ INFO] [1601538575.387566298]: Expected 1 Byte(s) got 0, occured 0 times                            
[ INFO] [1601538575.387581643]: std::exception, occured 0 times                                       
[ INFO] [1601538576.450039621]: Comm time dt: 0.001063, freq: 941.145484, errors: 0, error rate: 0.00%
[ INFO] [1601538576.450101267]: Expected 1 Byte(s) got 0, occured 0 times                            
[ INFO] [1601538576.450116695]: std::exception, occured 0 times                                       
[ INFO] [1601538577.555609399]: Comm time dt: 0.001106, freq: 904.517271, errors: 1, error rate: 0.10%
[ INFO] [1601538577.555671551]: Expected 1 Byte(s) got 0, occured 1 times                            
[ INFO] [1601538577.555691719]: std::exception, occured 0 times                                       
[ INFO] [1601538578.613995893]: Comm time dt: 0.001058, freq: 944.824532, errors: 0, error rate: 0.00%
[ INFO] [1601538578.614048956]: Expected 1 Byte(s) got 0, occured 0 times                            
[ INFO] [1601538578.614064618]: std::exception, occured 0 times                                       
[ INFO] [1601538579.687532013]: Comm time dt: 0.001074, freq: 931.504165, errors: 0, error rate: 0.00%
[ INFO] [1601538579.687590877]: Expected 1 Byte(s) got 0, occured 0 times                            
[ INFO] [1601538579.687613109]: std::exception, occured 0 times                                       
```

After:
```
[ INFO] [1601538544.118066231]: Comm time dt: 0.001265, freq: 790.275380, errors: 14, error rate: 1.40%
[ INFO] [1601538544.118129433]: Expected 1 Byte(s) got 0, occured 14 times                            
[ INFO] [1601538544.118145100]: std::exception, occured 1 times                                        
[ INFO] [1601538545.363324218]: Comm time dt: 0.001245, freq: 803.039438, errors: 14, error rate: 1.40%
[ INFO] [1601538545.363390532]: Expected 1 Byte(s) got 0, occured 14 times                            
[ INFO] [1601538545.363430219]: std::exception, occured 0 times                                        
[ INFO] [1601538546.577911849]: Comm time dt: 0.001215, freq: 823.324648, errors: 7, error rate: 0.70% 
[ INFO] [1601538546.577968530]: Expected 1 Byte(s) got 0, occured 7 times                             
[ INFO] [1601538546.578001358]: std::exception, occured 0 times                                        
[ INFO] [1601538547.803263941]: Comm time dt: 0.001225, freq: 816.093233, errors: 15, error rate: 1.50%
[ INFO] [1601538547.803321719]: Expected 1 Byte(s) got 0, occured 15 times                            
[ INFO] [1601538547.803338930]: std::exception, occured 0 times                                        
[ INFO] [1601538549.008076665]: Comm time dt: 0.001205, freq: 830.005356, errors: 12, error rate: 1.20%
[ INFO] [1601538549.008186700]: Expected 1 Byte(s) got 0, occured 12 times                            
[ INFO] [1601538549.008203645]: std::exception, occured 0 times                                        
```

**Comment by gbalke** [comment]
This returned operation to nominal levels. https://github.com/BetzDrive/bldc-controller/tree/v2_2-comms_fix

**Comment by gbalke** [comment]
I believe this has been resolved by #45.
